File: 94_Binary_Tree_Inorder_Traversal/javascript_solutions/StackSolution.ts
function StackSolution(root: TreeNode | null): number[] {
    if (root == null) return [];

    const stack: TreeNode[] = [];
    const result: number[] = [];

    while(root || stack.length){
        while(root){
            stack.push(root);
            root = root.left;
        }

        const top = stack.pop();
        if(!top) throw new Error("Data error");
        result.push(top.val);
        root = top.right;
    }

    return result;
};

File: 94_Binary_Tree_Inorder_Traversal/java_solutions/StackSolution.java
import java.util.ArrayList;
import java.util.List;
import java.util.Stack;

public class StackSolution {
    public List<Integer> inorderTraversal(TreeNode root) {
        List<Integer> result = new ArrayList<>();
        if(root == null) return result;

        Stack<TreeNode> stack = new Stack<>();

        while(root != null || !stack.empty()){
            while(root != null){
                stack.add(root);
                root = root.left;
            }

            TreeNode top = stack.pop();
            result.add(top.val);
            root = top.right;
        }

        return result;
    }
}
